Concordium DID Method (#39)
Concordium/concordium-update-proposals: source/ID/concordium-did.rst

We distinguish the following types of DIDs:

- **Account DID** refer to accounts on the Concordium blockchain.
- **Smart Contract DID** refer to smart contracts instances on the Concordium blockchain.
- **Public Key DID** refers to a subject that knows the corresponding secret key.

Status of This Document
=======================

This is a draft document and may be updated.

[Add more legal text here?]

Specification
=============

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in :rfc:`2119`.

Concordium DID Identifier
=========================

Method Name
-----------

The Concordium DID method name is ``ccd``.
A DID identifier that uses this method must begin with the following prefix: ``did:ccd``.
Per DID specification [DID] this string must be in lowercase.

Identifier Syntax
-----------------

Concordium DID identifiers are defined by the following ABNF_:

.. code-block:: ABNF

  ccd-did = prefix ":" *1(network ":") (acctype / pkctype / scitype)
  prefix  = %s"did:ccd"

>   network = "testnet" / "mainnet"

Reviewer: Not clear what exactly network identifier is and it would be nice to define it somewhere (for example, for "pkctype: did-s, I have no idea what network identifier would mean.)

Author: Yes, good point, I think we should make a separate case for `pkctype`, I agree that it's not clear what network means in this context.

Author: As we agreed, since the network is optional, we keep the syntax as it is, but add a note that if the network is present, it could affect the generated DID document in the future (e.g. different verification method).

  punctuation = "!" / DQUOTE / "#" / "$" / "%" / "&" / "'" / "(" /
                ")" / "*" / "+" / "," / "-" / "." / "/" / ";" /
                "<" / "=" / ">" / "?" / "@" / "[" / "\" / "]" /
                "^" / "_" / "`" / "{" / "|" / "}" / "~"

See the details about *dereferencing* a smart contract instance reference in :ref:`sci-reference`.

Examples
--------

An account on testnet:

``did:ccd:testnet:acc:3ZFGxLtnUUSJGW2WqjMh1DDjxyq5rnytCwkSqxFTpsWSFdQnNn``

A smart contract instance on the default network (``mainnet``):

``did:ccd:sci:12:0``

A public key related to ``mainnet``:

``did:ccd:mainnet:pkc:0c7f4421e44a4385850b883e3bbf098f5a9853ef6f1a862c2ce2856381b5f5e3``

A smart contract instance with the ``getKeys`` entrypoint that does not take any parameters

``did:ccd:sci:321/getKeys``

A smart contract instance with the ``getCredentials`` entrypoint taking a parameter

``did:ccd:sci:123/getCredentials?parameter=QWWfHFtjPSLJdrz1beyVPzFmgS2ECbJVe1k2LXPwUGMZRjjpL``

Concordium DID Documents
========================

.. TODO add formal DID documents

Account DID
-----------

The Account DID Document MUST contain the following data:

> - ``@context``

Reviewer: ```suggestion
- ``@context`` - the attribute that expresses context information.
```

- ``id`` - the DID of the account.
- ``accountCredentials`` - a list of account holders represented by their account credentials ``credential-i`` for ``i = 1..M``, where ``M`` is the number of credentials.
  The credentials contain several (at least one) signature verification keys ``key-i-j`` for ``j = 1..N_i``, where ``N_i`` is the number of keys, which can be different for each credential.

>   For each credential:

Reviewer: Itemization is a bit unclear here (only the second item has a bullet). Also, "verificationMethod" should be a property for the DID; so, defining it as a "For each credential:" is a bit informal here

Author: I'll make it into a nested `vefiricationMethod`

  - ``verificationMethod`` - uses a threshold verification scheme that specifies public keys ``key-i-j`` and a signature threshold ``R_j``.
- ``verificationMethod`` - account's verification method; it is again a threshold scheme requiring at least ``T`` credentials to sign.

The document MAY include any other public data of a Concordium account.

.. code-block:: json

  {
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "Concordium DID URI"
    ],
    "id": "did:ccd:NET:acc:ADDR",
    "accountCredentials": [
      {

>         "verificationMethod": [

Reviewer: Would be nice to have the formal definition of "verificationMethod" (i.e., defining its properties e.g., id, controller, type, etc)

Author: I added a reference to the w3c did standard

        ]
      }
    ],
    "verificationMethod": [
      {
        "id": "did:ccd:NET:acc:ADDR#acc-1",
        "controller": "did:ccd:NET:acc:ADDR",
        "type": "VerifiableCondition2021",
        "threshold": "T",
        "conditionThreshold": [
          "#credential-1",
          "...",
          "#credential-M"
        ]
      }
    ],
    "authentication": [
      "#acc-1"
    ]
  }

See the details about the verification method extension in :ref:`concordium-did-verification-method`.

Smart Contract Instance DID
---------------------------

The Smart Contract Instance DID Document MUST contain the following data:

- ``@context``
- ``id`` - the DID of the smart contract instance.
- ``owner`` - a DID of an account that initialized the contract instance.

The document MAY include any other public data of a smart contract instance.

.. code-block:: json

  {
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "Concordium DID URI"
    ],
    "id": "did:ccd:sci:IND:SUBIND",
    "owner": "did:ccd:NET:acc:ADDR"
  }

Where ``IND`` and ``SUBIND`` are the contract index and subindex.
``NET`` and ``ADDR`` correspond to the network and to the owner's account address.

> Public Key Cryptography DID

Author: Done.

The entrypoint is considered a *view*: no state changes are persisted, only the result of the invocation is returned to the caller.
The result of the invocation is the return value produced by the contract or an error, if the invocation failed.
The return value is in the JSON format corresponding to the embedded smart contract schema.

.. TODO should the binary return values be allowed? What if the contract doesn't have an embedded schema?

From the command line, ``concordium-client`` allows for invoking a smart contract instance in the following way:

.. code-block::

  $concordium-client contract invoke IND --entrypoint EP --energy 3000000 --parameter-json param.json

The base58 encoding of the ``param.json`` file corresponds to ``PAR``.

See the details in the `gRPC v2 documentation`_.

.. seealso::

  `Dereferencing a DID URL`_ in the W3C Credentials Community Group draft report.

.. TODO: write about binary vs JSON data

Public Key Cryptography DID
^^^^^^^^^^^^^^^^^^^^^^^^^^^

The DID document corresponding to a DID of the form

``did:ccd:NET:pkc:PK``

can be constructed directly from the DID without any lookup necessary.

> .. code-block:: json

Reviewer: The code block doesn't seem very relevant. So, I would just remove it (unless it's giving some information about pkc-DID document that I cannot see?).

Author: Removed and added a note about the network

  {
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "Concordium DID URI"
    ],
    "id": "did:ccd:NET:pkc:PK",
    "verificationMethod": [
      {
        "id": "did:ccd:pkc:PK#key-0",
        "type": "Ed25519VerificationKey2020",
        "controller": "did:ccd:NET:pkc:PK",
        "publicKeyMultibase": "PK"
      }
    ],
    "authentication": [
      {
        "did:ccd:NET:pkc:PK#key-0"
      }
    ]
  }

Update
------

At this time Concordium does not support the update of DID documents.

.. TODO Technically the account based DIDs are updateable, add something about it?

Deactivate
----------

At this time Concordium does not support deactivation of DID documents.

Appendices
==========

.. _concordium-did-verification-method:

Threshold Verification Method
-----------------------------

> The threshold verification method usen in Concordium DID Documents is based on a `ConditionalProof verification method <https://w3c-ccg.github.io/verifiable-conditions/>`_.

Reviewer: ```suggestion
The threshold verification method used in Concordium DID Documents is based on a `ConditionalProof verification method <https://w3c-ccg.github.io/verifiable-conditions/>`_.
```

This is a new type of verification method under development.
``ConditionalProof`` features several extensions such as logical operations (``and``, ``or``), threshold and weighted threshold.
Note that the method is not yet a W3C standard and currently has a *draft* status.

> The example below shows the ``2-out-of-3`` signature verifcation method.

Reviewer: ```suggestion
The example below shows the ``2-out-of-3`` signature verification method.
```

We distinguish the following types of DIDs:

- **Account DID** refer to accounts on the Concordium blockchain.
- **Smart Contract DID** refer to smart contracts instances on the Concordium blockchain.
- **Public Key DID** refers to a subject that knows the corresponding secret key.

Status of This Document
=======================

This is a draft document and may be updated.

[Add more legal text here?]

Specification
=============

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in :rfc:`2119`.

Concordium DID Identifier
=========================

Method Name
-----------

The Concordium DID method name is ``ccd``.
A DID identifier that uses this method must begin with the following prefix: ``did:ccd``.
Per DID specification [DID] this string must be in lowercase.

Identifier Syntax
-----------------

Concordium DID identifiers are defined by the following ABNF_:

.. code-block:: ABNF

  ccd-did = prefix ":" *1(network ":") (acctype / pkctype / scitype)
  prefix  = %s"did:ccd"
  network = "testnet" / "mainnet"
  acctype = "acc:" 50(base58char)

>   pkctype = "pkc:" 64(base58char)

Author: I think we took some of this from DID methods for other chains. We should just use what is the most meaningful for us. The same applies to other cases related to encoding.

>   pkctype = "pkc:" 64(base58char)

Author: Account addresses are represented as base58, and let's use base16 for public keys.

  scitype = "sci:" index *1(“:” subindex)
  index = 1*DIGIT
  subindex = 1*DIGIT
  base58char = "1" / "2" / "3" / "4" / "5" / "6" / "7" / "8" / "9" /
               "A" / "B" / "C" / "D" / "E" / "F" / "G" / "H" / "J" /
               "K" / "L" / "M" / "N" / "P" / "Q" / "R" / "S" / "T" /
               "U" / "V" / "W" / "X" / "Y" / "Z" / "a" / "b" / "c" /
               "d" / "e" / "f" / "g" / "h" / "i" / "j" / "k" / "m" /
               "n" / "o" / "p" / "q" / "r" / "s" / "t" / "u" / "v" /
               "w" / "x" / "y" / "z"

.. note::
    The network part of a CCD DID is optional.
    *If omitted the network is assumed to be "mainnet" by default*.

.. note::
    The ``subindex`` part of the smart contract instance ``sci`` is optional.
    *If omitted the subindex is assumed to be `0` by default.*

Smart Contract Instance Reference Syntax
----------------------------------------

A DID reference is a DID with additional data, such as query parameters.
Concordium smart contract entrypoints can be addressed using DID references.

.. code-block:: ABNF

>   ccd-sc-ref = "/" entrypoint *1("?" param-name "=" 1*(base58char))

Author: @abizjak What do you think would be a better choice here: `base64` or `base16`?

  param-name = "parameter_json" / "parameter_binary"
  entrypoint = 1*(ALPHA / DIGIT / punctuation)
  punctuation = "!" / DQUOTE / "#" / "$" / "%" / "&" / "'" / "(" /
                ")" / "*" / "+" / "," / "-" / "." / "/" / ";" /
                "<" / "=" / ">" / "?" / "@" / "[" / "\" / "]" /
                "^" / "_" / "`" / "{" / "|" / "}" / "~"

See the details about *dereferencing* a smart contract instance reference in :ref:`sci-reference`.

Examples
--------

An account on testnet:

``did:ccd:testnet:acc:3ZFGxLtnUUSJGW2WqjMh1DDjxyq5rnytCwkSqxFTpsWSFdQnNn``

A smart contract instance on the default network (``mainnet``):

``did:ccd:sci:12:0``

A public key:

``did:ccd:pkc:0c7f4421e44a4385850b883e3bbf098f5a9853ef6f1a862c2ce2856381b5f5e3``

A smart contract instance with the ``getKeys`` entrypoint that does not take any parameters

``did:ccd:sci:321/getKeys``

A smart contract instance with the ``getCredentials`` entrypoint taking a parameter

> ``did:ccd:sci:123/getCredentials?parameter_json=QWWfHFtjPSLJdrz1beyVPzFmgS2ECbJVe1k2LXPwUGMZRjjpL``

Reviewer 2: Do we really want to support JSON for this? What is the value in this as opposed to just the parameter in binary that is passed directly to the contract?

Author: Let's use binary data.

Account DID
-----------

The goal of the Account DID Document is to provide information about the account authentication data, including a possibility to reference particular pieces of data, such as public keys.
In order to do that, it specifies a `DID verification method <did-vefication-method_>`_ that reflects the account authentication data: public keys grouped into credentials.

The Account DID Document MUST contain the following data:

- ``@context`` - the attribute that expresses context information.
- ``id`` - the DID of the account.
- ``verificationMethod`` - the account's verification method.
  It is a nested :ref:`threshold scheme <concordium-did-verification-method>` requiring at ``T`` out of ``M`` credentials to sign; each credential uses its own threshold scheme requiring ``R_i`` out of ``N_i`` keys to sign, where ``i = 1..M``.and ``j = 1..N_i``.
  The credentials are identified by a `DID fragment`_ ``#credential-i``, and the keys in each credentials by ``#key-j-i`` where ``i = 1..M`` and ``j = 1..N_i``.
- ``authentication`` - authentication method for the account.

The document MAY include any other public data of a Concordium account.

.. note::

>   A `DID fragment`_ allows for referencing a particular credential, or a key in the Account DID Document.

Reviewer 2: Does it mean we need to support this in the resolver? Or not?

                "threshold": "R_1",
                "conditionThreshold": [
                  {
                    "id": "did:ccd:NET:acc:ADDR#key-1-1",
                    "type": "Ed25519VerificationKey2020",
                    "controller": "did:ccd:NET:acc:ADDR#credential-1",
                    "publicKeyMultibase": "XX"
                  },
                  "...",
                  {
                    "id": "did:ccd:NET:acc:ADDR#key-N_1-1",
                    "type": "Ed25519VerificationKey2020",
                    "controller": "did:ccd:NET:acc:ADDR#credential-1",
                    "publicKeyMultibase": "YY"
                  }
                ]
              }
            ]
          },
          "...",
          {
            "verificationMethod": [
              {
                "id": "did:ccd:NET:acc:ADDR#credential-M",
                "controller": "did:ccd:NET:acc:ADDR",
                "type": "VerifiableCondition2021",
                "threshold": "N",
                "conditionThreshold": [
                  {
                    "id": "did:ccd:NET:acc:ADDR#key-1-M",
                    "type": "Ed25519VerificationKey2020",

>                     "controller": "did:ccd:NET:acc:ADDR#credential-M",

Reviewer 2: Why is this a controller? According to the spec this looks to be
```
An entity that has the capability to make changes to a [DID document](https://www.w3.org/TR/did-core/#dfn-did-documents). 
```

But why does credential-M have this capability? On its own it does not.

Author: `credential-M` is the controller for the public key that belongs to this credential. Maybe I don't fully understand the idea. I thought that each credential (as part of an account) controls its keys.

Reviewer 2: Well it does in the sense that the keys cannot be changed without the credential signing. But that is on its own not enough. So in that sense, if the credential is part of a multi-credential account it has to have cooperation of others to change the keys.

Author: Ok, then is makes sense to make the account `did:ccd:NET:acc:ADDR` to be the controller.

                    "publicKeyMultibase": "VV"
                  },
                  "...",
                  {
                    "id": "did:ccd:NET:acc:ADDR#key-N_M-M",
                    "type": "Ed25519VerificationKey2020",
                    "controller": "did:ccd:NET:acc:ADDR#credential-M",
                    "publicKeyMultibase": "ZZ"
                  }
                ]
              }
            ]
          }
        ]
      }
    ],
    "authentication": [
      "#acc-1"
    ]
  }

Smart Contract Instance DID
---------------------------

The goal of the Smart Contract Instance DID is to provide meta-data about the contract instance.
At the moment, the main piece of meta-data is the Concordium account that send the initialization transaction.

The Smart Contract Instance DID Document MUST contain the following data:

- ``@context`` - the attribute that expresses context information.
- ``id`` - the DID of the smart contract instance.

> - ``owner`` - a DID of an account that initialized the contract instance represented as a JSON object containing fields ``id`` and ``account``.

Reviewer 2: I don't think we should include this. If we do it should say "creator" instead.

Author: One reason to include this is that the issuers are account addresses that created an instance of a credential registry. I think there should be a way of referring to the contract owner.  We could use `creator` if this makes more sense.

    "verificationMethod": [
      {
        "id": "did:ccd:pkc:XX#key-0",
        "type": "Ed25519VerificationKey2020",
        "controller": "did:ccd:NET:pkc:PK",
        "publicKeyMultibase": "XX"
      }
    ],
    "authentication": [
      {
        "did:ccd:pkc:XX#key-0"
      }
    ]
  }

Concordium DID Operations
=========================

Concordium DIDs are managed on the Concordium blockchain.

Create
------

Account DID
^^^^^^^^^^^

An account DID can be created by `opening an account <concordium-accounts_>`_ on the ``NET`` blockchain.
The resulting DID is ``did:ccd:NET:acc:ADDR`` where ``ADDR`` is the base58 encoded account address.

Smart Contract Instance DID
^^^^^^^^^^^^^^^^^^^^^^^^^^^

A smart contract instance DID can be created by `deploying a smart contract module <deploy-module_>`_ and `initializing a smart contract instance <initialize-contract-instance_>`_ on the ``NET`` blockchain.
The resulting DID is ``did:ccd:NET:sci:IND:SUBIND`` where ``IND``, ``SUBIND`` are the index and the subindex of the instance.

Public Key Cryptography DID
^^^^^^^^^^^^^^^^^^^^^^^^^^^

A public key cryptography DID can be created by generating a fresh Ed25519 key pair.

> The resulting DID is ``did:ccd:NET:pkc:PK`` where ``PK`` is the base58 encoded public key.

Reviewer 2: I strongly recommend base16 here.

Author: Yes, let's use base16.

See the details in the `gRPC v2 documentation`_.

From the command line, ``concordium-client`` allows for retrieving the data in the following way:

.. code-block:: console

  $concordium-client contract show IND

.. TODO update, once we have a DID resolver

.. _sci-reference:

Smart Contract Instance Reference
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*Dereferencing* the smart contract DID reference invokes the specified entrypoint.

Dereferencing a DID reference of the form

``did:ccd:NET:sci:IND:SUBIND/EP?parameter_json=PAR[&format=FMT]``

can be done by using the gRPC interface command ``InvokeInstance``.
The entrypoint is considered a *view*: no state changes are persisted, only the result of the invocation is returned to the caller.
The parameter ``PAR`` is passed to the entrypoint.

The result of the invocation is the return value produced by the contract or an error, if the invocation failed.
The optional query parameter ``format=FMT`` specifies the formatting of the return value.
If not specified, the return value is in the JSON format corresponding to the embedded smart contract schema.

> If specified, the ``FMT`` value can be the following:

Reviewer 2: Is it an error if this is specified and the schema is not there?

I would generally prefer not to rely on any schemas. They are optional.

Author: After discussing the formatting of return results with @abizjak , the proposed solution is the following. Instead of having the `format` parameter, we introduce the `standard` parameter, which refers to a CIS standard associated with the smart contract. For each standard, we define how the return data should be rendered. Specifically for the smart contract registries, we specify that the entrypoints that return public keys a rendered as DID documents with the appropriate  `verificationMethod`.

- ``ed25519-pk`` - the entrypoint return value is expected to be a string, e.g. ``"XX"``.
  The output is a DID document with a verification method based on a public key ``XX``:

  .. code-block:: json

    {
      "@context": [
        "https://www.w3.org/ns/did/v1",
        "Concordium DID URI"
      ],
      "id": "did:ccd:NET:sci:IND:SUBIND/EP?parameter_json=PAR&format=ed25519-pk",
      "verificationMethod": [
        {
          "id": "did:ccd:NET:sci:IND:SUBIND/EP?parameter_json=PAR&format=ed25519-pk#key-0",
          "type": "Ed25519VerificationKey2020",
          "publicKeyMultibase": "XX"
        }
      ]
    }

> - ``ed25519-pk-list`` - the entrypoint return value is expected to be a list of strings, e.g. ``["XX", "YY", "ZZ"]``.

Reviewer 2: Why is the entrypoint not returning a list of byte arrays instead?

.. _concordium_did:

==================================
Concordium DID Method Version 1.0
==================================

About
=====

The Concordium DID method specification conforms to the requirements specified in `Decentralized Identifiers (DIDs) v1.0 <w3c-did-core-v1.0_>`_ [DID] published by the W3C Credentials Community Group.
For more information about DIDs and DID method specifications, please see the `DID Primer`_ and `DID Spec`_.

Abstract
=========

Concordium is a layer-1 blockchain with identity built into the core protocol.
We distinguish the following types of DIDs:

> - **Account DID** refer to accounts on the Concordium blockchain.

Reviewer 2: If the idea is that we use these uniformly to refer to credentials, then I think we need one to refer to the existing identity credentials. Referring to an account is not the right thing since it can have more than one credential.

In ID 2.0 setup the proof is associated with a credential registration ID.

We can add this part to the path and use the account DID, but we need to then define that. It seems like this solution with fragments
```
did:ccd:NET:acc:ADDR#credential-1
```
is worse than just being able to refer to a credential via its ID.

Author: It's an interesting point.
I don't know the exact use case for account DIDs. I imagine it's for off-chain verifiable credentials signed by the issuer (with the issuer's account keys). In this case, there might be various options, I believe.  It could be an account that sings according to the whole multi-sig structure, an individual credential, or a single key.
We could add a credential id to the fragment, instead of using indices like `credential-1`. But it might be more convenient to have a separate DID  type for that. Especially if the credential IDs are unique (are they?). I thought about credentials as part of the account structure, not as independent entities, but I could be wrong.

Maybe @tschudid can also add something to this discussion?

Reviewer 2: I am thinking pragmatically. If the dapp asks for a proof that involves the user's ID credentials, issued by an identity provider, then we need to put some metadata into the proof as to what on-chain credential this is referring to. At the moment this is just a credential registration ID directly. If we want to have a uniform way of identifying, this should now be a DID. How should it look? An account can have multiple credentials from different identities.

Credential registration IDs are unique. That is enforced by the chain.

Author: Yes, I agree with that. My real question was whether a ruse credential can be fetched by it's ID directly, or one still neds to know the account address. If the account address is still needed, then fragments could be a way to go.
For example
`did:ccd:acc:4LxatD2pbmKe2zJWNF1QsSVRWPVgnfLnPPyfxVEor4pohZrKzc#8113641a210da36a9ffae6e6095b9cf486ca9b44fa059aa74577b0a1c0f7052d8e71168beccf299d767f3961b33aaae2`
In this case the fragment is the value of the `credId` field that I get through gRPC.

Then the account DID document will contain the appropriate fragment
```
{
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "Concordium DID URI"
    ],
    "id": "did:ccd:testnet:acc:4LxatD2pbmKe2zJWNF1QsSVRWPVgnfLnPPyfxVEor4pohZrKzc",
    "verificationMethod": [
      {
        "id": "did:ccd:testnet:acc:4LxatD2pbmKe2zJWNF1QsSVRWPVgnfLnPPyfxVEor4pohZrKzc#acc-1",
        "controller": "did:ccd:NET:acc:ADDR",
        "type": "VerifiableCondition2021",
        "blockchainAccountId": "ADDR",
        "threshold": "T",
        "conditionThreshold": [
          {
            "verificationMethod": [
              {
                "id": "did:ccd:acc:4LxatD2pbmKe2zJWNF1QsSVRWPVgnfLnPPyfxVEor4pohZrKzc#8113641a210da36a9ffae6e6095b9cf486ca9b44fa059aa74577b0a1c0f7052d8e71168beccf299d767f3961b33aaae2",
                "controller": "did:ccd:NET:acc:ADDR",
                "type": "VerifiableCondition2021",
                "threshold": "R_1",
                "conditionThreshold": [
                  {
                    "id": "did:ccd:NET:acc:ADDR#key-1-1",
                    "type": "Ed25519VerificationKey2020",
                    "controller": "did:ccd:testnet:acc:4LxatD2pbmKe2zJWNF1QsSVRWPVgnfLnPPyfxVEor4pohZrKzc",
                    "publicKeyMultibase": "fXX"
                  },
                  "...",
                  {
                    "id": "did:ccd:NET:acc:ADDR#key-N_1-1",
                    "type": "Ed25519VerificationKey2020",
                    "controller": "did:ccd:testnet:acc:4LxatD2pbmKe2zJWNF1QsSVRWPVgnfLnPPyfxVEor4pohZrKzc",
                    "publicKeyMultibase": "fYY"
                  }
                ]
              }
            ]
          },
          "...",
            ]
          }
        ]
      }
    ],
 ...
  }
```

Otherwise, if it's possible to resolve a Concordium user credential directly, just by it's ID, we should introduce a new DID type.

Something like

```
did:ccd:NET:cred:8113641a210da36a9ffae6e6095b9cf486ca9b44fa059aa74577b0a1c0f7052d8e71168beccf299d767f3961b33aaae2
```

Reviewer 2: You can look up via a credential ID directly. What you get is an account though, an account that has or has had the credential associated with it.

- **Smart Contract DID** refer to smart contracts instances on the Concordium blockchain.

> - **Public Key DID** refers to a subject that knows the corresponding secret key.

Reviewer 2: This is missing the IDP type.

  network = "testnet" / "mainnet"
  acctype = "acc:" 50(base58char)
  pkctype = "pkc:" 64(base16char)
  scitype = "sci:" index *1(“:” subindex)
  idptype = "idp:" index
  credtype = "cred:" 96(base16char)
  index = 1*DIGIT
  subindex = 1*DIGIT
  base16char = HEXDIG
  base58char = "1" / "2" / "3" / "4" / "5" / "6" / "7" / "8" / "9" /
               "A" / "B" / "C" / "D" / "E" / "F" / "G" / "H" / "J" /
               "K" / "L" / "M" / "N" / "P" / "Q" / "R" / "S" / "T" /
               "U" / "V" / "W" / "X" / "Y" / "Z"

.. note::
    In the ABNF_ grammar, string literals are case-insensitive, unless explicitly specified using the ``%s`` prefix.

.. note::
    The network part of a CCD DID is optional.
    *If omitted the network is assumed to be "mainnet" by default*.

.. note::
    The ``subindex`` part of the smart contract instance ``sci`` is optional.
    *If omitted the subindex is assumed to be `0` by default.*

Smart Contract Instance Reference Syntax
----------------------------------------

A DID reference is a DID with additional data, such as a path and query parameters.
Concordium smart contract entrypoints can be addressed using DID references.

.. code-block:: ABNF

>   ccd-sc-ref = "/" entrypoint *1("/" parameter) *1("?" "standard" "=" standard)

Author: Yes, it should be removed, I'll do that.

  parameter = base16char
  entrypoint = 1*(ALPHA / DIGIT / punctuation)

>   standard = "CIS" "-" 1*(DIGIT)

Author: Will be removed.

The Account DID Document MUST contain the following data:

- ``id`` - the DID of the account.
- ``verificationMethod`` - the account's verification method.
  It is a nested :ref:`threshold scheme <concordium-did-verification-method>` requiring at ``T`` out of ``M`` credentials to sign; each credential uses its own threshold scheme requiring ``R_i`` out of ``N_i`` keys to sign, where ``i = 1..M``.and ``j = 1..N_i``.
  The credentials are identified by a `DID fragment`_ ``#credential-i``, and the keys in each credentials by ``#key-j-i`` where ``i = 1..M`` and ``j = 1..N_i``.
- ``authentication`` - authentication method for the account.

The document MAY include any other public data of a Concordium account.

.. note::

  A `DID fragment`_ allows for referencing a particular credential, or a key in the Account DID Document.
  The fragment is used to locate the (unique) JSON object by matching the DID URL with the object's ``id`` property.

.. seealso::

  `Dereferencing a DID URL`_ in the W3C Credentials Community Group draft report.

.. code-block:: json

  {
    "id": "did:ccd:NET:acc:ADDR",
    "verificationMethod": [
      {

>         "id": "did:ccd:NET:acc:ADDR#acc-1",

Author: The fragment `#acc-1` is used to identify a verification method for the account. It could be any string; `acc-1` is confusing, so I'll change it to, e.g. `acc-vm`.  It is used as a relative reference in `authentication`. That is, this verification method is used to authenticate operations for the account.

>         "id": "did:ccd:NET:acc:ADDR#acc-1",

Author: Fiexd here and in the credential DID document as well.

      { "id": "did:ccd:sci:IND:SUBIND#entrypoint-revocationKey",
        "name": "revocationKey"
      }
    ]
  }

Where ``IND`` and ``SUBIND`` are the contract index and subindex.
``NET`` and ``ADDR`` correspond to the network and to the owner's account address.

.. _concordium-did-pkc:

Public Key Cryptography DID
---------------------------

The goal of the Public Key Cryptography DID is to represent a public key and the corresponding signature verification method.

The Public Key Cryptography DID Document MUST contain the following data:

- ``id`` - the DID of the public key.
- ``verificationMethod`` - specifies a `DID verification method <did-vefication-method_>`_ for verifying a signature corresponding to the public key.
- ``authentication`` - authentication method for the key.

.. code-block:: json

  {
    "id": "did:ccd:pkc:XX",
    "verificationMethod": [
      {
        "id": "did:ccd:pkc:XX#key-0",
        "type": "Ed25519VerificationKey2020",
        "controller": "did:ccd:NET:pkc:PK",
        "publicKeyMultibase": "fXX"
      }
    ],
    "authentication": [
      {
        "did:ccd:pkc:XX#key-0"
      }
    ]
  }

Identity Provider DID
---------------------

> The goal of the Identity Provider DID is identify a Concodrium identity provider (IDP).

Reviewer 2: ```suggestion
The goal of the Identity Provider DID is identify a Concordium identity provider (IDP).
```

An identity provider is an organization, approved by Concordium, that performs off-chain identification of users.
IDPs are used in the account creation process to issue an identity.

> IDP DIDs can represent an issuer of a verifiable credential.

Reviewer 2: I recommend spelling out identity providers.

I think acronyms in this case are needless

Author: Fixed.

The Identity Provider DID Document MUST contain the following data:
- ``id`` - the DID of the IDP.
- ``name`` - the IDP name.
- ``url`` - A link to more information about the IDP.
- ``description`` - A free form description the IDP.

> - ``verificationMethod`` - specifies a `DID verification method <did-vefication-method_>`_ for verifying a signature corresponding to the public key.

Reviewer 2: This does't render as an itemized lits. It's all inline.

Author: Fixed.

.. code-block:: json

  {
    "id": "did:ccd:testnet:idp:3",
    "name": "Digital Trust Solutions TestNet",
    "url": "https://www.digitaltrustsolutions.nl",

>     "description": "Identity verified by Digital Trust Solutions on behalf of Concordium",

Reviewer 2: We should probably use the test identity provider in this. Not DTS

Author: Fixed.

    "verificationMethod": [
      {
        "id": "did:ccd:testnet:idp:3#cdi-key",
        "type": "Ed25519VerificationKey2020",
        "controller": "did:ccd:NET:pkc:PK",
        "publicKeyMultibase": "fXX"
      }
    ]
  }

Concordium DID Operations
=========================

Concordium DIDs are managed on the Concordium blockchain.

Create
------

Account DID
^^^^^^^^^^^

An account DID can be created by `opening an account <concordium-accounts_>`_ on the ``NET`` blockchain.
The resulting DID is ``did:ccd:NET:acc:ADDR`` where ``ADDR`` is the base58 encoded account address.

Concordium Credential DID
^^^^^^^^^^^^^^^^^^^^^^^^^

> A Concordium Credential DID is created as part of the account opening process.

Reviewer 2: Or as part of adding acredential onto an existing account.

Author: Fixed.

These DIDs are not registered on the blockchain.

Identity Provider DID
^^^^^^^^^^^^^^^^^^^^^

Identity providers can be added as a `chain update <https://docs.rs/concordium_base/1.2.0/concordium_base/updates/index.html>`_ transaction of type `UpdateAddIdentityProvider <https://docs.rs/concordium_base/1.2.0/concordium_base/updates/enum.UpdateType.html#variant.UpdateAddIdentityProvider>`_.

Read
----

Account DID
^^^^^^^^^^^

The DID document information for a DID of the form

``did:ccd:NET:acc:ADDR``

can be resolved by looking up an account with address ``ADDR`` on blockchain ``NET``.

Data required to construct the DID document can be acquired by using the gRPC interface command ``GetAccountInfo``.

See the details in the `gRPC v2 documentation`_.

From the command line, ``concordium-client`` allows to retrieve the data in the following way:

.. code-block:: console

    $concordium-client raw GetAccountInfo ADDR

.. TODO update, once we have a DID resolver

Concordium Credential DID
^^^^^^^^^^^^^^^^^^^^^^^^^

The DID document information for a DID of the form

``did:ccd:NET:cred:CRED``

can be resolved by looking up a credential with ID ``CRED`` on blockchain ``NET``.

> Data required to construct the DID document can be acquired by using the same gRPC interface command ``GetAccountInfo`` as for Concodrium account DIDs.

Reviewer 2: ```suggestion
Data required to construct the DID document can be acquired by using the same gRPC interface command ``GetAccountInfo`` as for Concordium account DIDs.
```

.. seealso::

  `Dereferencing a DID URL`_ in the W3C Credentials Community Group draft report.

.. TODO update, once we have a DID resolver

Public Key Cryptography DID
^^^^^^^^^^^^^^^^^^^^^^^^^^^

The DID document corresponding to a DID of the form

``did:ccd:NET:pkc:PK``

can be constructed directly from the DID without any lookup necessary.

.. note::

  The ``NET`` part is optional and currently there is no difference how the documents are generated for different networks.
  In the future, however, the ``vefiricationMethod`` as it specified in :ref:`concordium-did-pkc` might depend on the network.

Identity Provider DID
^^^^^^^^^^^^^^^^^^^^^

The DID document information for a DID of the form

``did:ccd:NET:idp:INDEX``

can be resolved by looking up an identity provider ``INDEX`` on blockchain ``NET``.

Data required to construct the DID document can be acquired by using the gRPC interface command ``GetIdentityProviders``.

See the details in the `gRPC v2 documentation`_.

From the command line, ``concordium-client`` allows to retrieve the data in the following way:

.. code-block:: console

    $concordium-client raw GetIdentityProviders

Update
------

> At this time Concordium does not support the update of DID documents.

Reviewer 2: Credential and account keys can be updated, not sure if that matters.

Author: I added some general info referencing Rust code documentation. Here and below for deactivation.

Deactivate
----------

> At this time Concordium does not support deactivation of DID documents.

Reviewer 2: Credentials can be removed. Not sure if this matters for this or not.

Security Considerations
=======================

> The ``did:ccd`` method is built on top the Concordium blockchain, a public permissionless DLT.

Reviewer 3: DL makes more sense

Authorization is performed by means of digital signature keys.
Leakage of private keys allows an attacker to take control.
Parties therefore MUST handle private keys with care.

Privacy Considerations
=======================

DIDs SHOULD be assumed to be pseudonyoums and public as they might be stored on the underlying DLT.
Correlation attacks MAY be possible if information assocciated to DIDs is published.
It is therefore NOT RECOMMENDED to reuse PKC DIDs.

Appendices
==========

.. _concordium-did-verification-method:

Threshold Verification Method
-----------------------------

The threshold verification method used in Concordium DID Documents is based on a `ConditionalProof verification method <https://w3c-ccg.github.io/verifiable-conditions/>`_.
This is a new type of verification method under development.
``ConditionalProof`` features several extensions such as logical operations (``and``, ``or``), threshold and weighted threshold.
Note that the method is not yet a W3C standard and currently has a *draft* status.

The example below shows the ``2-out-of-3`` signature verification method.

> It uses the ``ConditionalProof2022`` verification method.

Author: `2022` is a version of the `ConditionalProof` verification method, apparently. This is how it is described in this document https://w3c-ccg.github.io/verifiable-conditions/#the-conditionalproof-type

``did:ccd:mainnet:cred:9aa3641a212da36a9ffae6e6085b9cf486ca9b44fa059aa74565b0a1c0f7052d8e71168beccf299d767f3961b33aaae2``

A smart contract instance on the default network (``mainnet``):

``did:ccd:sci:12:0``

A public key:

``did:ccd:pkc:0c7f4421e44a4385850b883e3bbf098f5a9853ef6f1a862c2ce2856381b5f5e3``

A smart contract instance with the ``issuerKeys`` entrypoint that does not take any parameters

``did:ccd:sci:321/issuerKeys``

A smart contract instance with the ``credentialEntry`` entrypoint taking a parameter

``did:ccd:sci:123/credentialEntry/ee763364dc1a47d6aa4cc6bdb005e2b2``

Concordium DID Documents
========================

Account DID
-----------

The goal of the Account DID Document is to provide information about the account authentication data, including a possibility to reference particular pieces of data, such as public keys.
In order to do that, it specifies a `DID verification method <did-vefication-method_>`_ that reflects the account authentication data: public keys grouped into credentials.

The Account DID Document MUST contain the following data:

- ``id`` - the DID of the account.
- ``verificationMethod`` - the account's verification method.

>   It is a nested :ref:`threshold scheme <concordium-did-verification-method>` requiring at ``T`` out of ``M`` credentials to sign; each credential uses its own threshold scheme requiring ``R_i`` out of ``N_i`` keys to sign, where ``i = 1..M``.and ``j = 1..N_i``.

Reviewer 4: The "at" before T is confusing. Was that meant to be something else?

Author: Thanks for pointing this out; "at" should be removed :)

Author: Done.

- ``authentication`` - authentication method for the credential.

The document MAY include any other public data of a Concordium credential.

The following document defines a Concordium credential with ID ``CRED``.
The credential has ``N`` keys and uses a threshold signature scheme requiring ``T`` signatures.

.. code-block:: json

  {
    "id": "did:ccd:NET:cred:CRED",
    "verificationMethod": [
      {
        "did:ccd:NET:cred:CRED#credential-vm"
        "type": "VerifiableCondition2021",
        "threshold": "T",
        "conditionThreshold": [
          {
            "id": "did:ccd:NET:cred:CRED#key-1",
            "type": "Ed25519VerificationKey2020",
            "controller": "did:ccd:NET:cred:CRED",
            "publicKeyMultibase": "fXX"
          },
          "...",
          {
            "id": "did:ccd:NET:cred:CRED#key-N",
            "type": "Ed25519VerificationKey2020",
            "controller": "did:ccd:NET:cred:CRED",
            "publicKeyMultibase": "fYY"
          }
        ]
      }
    ],
    "authentication": [
      "#credential-vm"
    ]
  }

Smart Contract Instance DID
---------------------------

> The goal of the Smart Contract Instance DID is to provide meta-data about the contract instance.

Reviewer 4: ```suggestion
The goal of the Smart Contract Instance DID is to provide metadata about the contract instance.
```

At the moment, it contains an account address of the initialization transaction sender, and the list of the contract's entrypoints.

The Smart Contract Instance DID Document MUST contain the following data:

- ``id`` - the DID of the smart contract instance.
- ``creator`` - a DID of an account that initialized the contract instance represented as a JSON object containing fields ``id`` and ``account``.
- ``entrypoints`` - a list on the contract's entrypoints. Each entrypoint is an object containing fields ``id`` and ``name``.

The document MAY include any other public data of a smart contract instance.

.. code-block:: json

  {
    "id": "did:ccd:sci:IND:SUBIND",
    "owner": {
      "id": "did:ccd:sci:IND:SUBIND#creator",
      "account": "did:ccd:NET:acc:ADDR"
    }
    "entrypoints": [
      { "id": "did:ccd:sci:IND:SUBIND#entrypoint-issuerKeys",
        "name": "issuerKeys"
      },
      { "id": "did:ccd:sci:IND:SUBIND#entrypoint-revocationKey",
        "name": "revocationKey"
      }
    ]
  }

Where ``IND`` and ``SUBIND`` are the contract index and subindex.

> ``NET`` and ``ADDR`` correspond to the network and to the owner's account address.

Reviewer 4: ```suggestion
``NET`` and ``ADDR`` correspond to the network and to the owner's account address respectively.
```

.. _concordium-did-pkc:

Public Key Cryptography DID
---------------------------

The goal of the Public Key Cryptography DID is to represent a public key and the corresponding signature verification method.

The Public Key Cryptography DID Document MUST contain the following data:

- ``id`` - the DID of the public key.
- ``verificationMethod`` - specifies a `DID verification method <did-vefication-method_>`_ for verifying a signature corresponding to the public key.
- ``authentication`` - authentication method for the key.

.. code-block:: json

  {
    "id": "did:ccd:pkc:XX",
    "verificationMethod": [
      {
        "id": "did:ccd:pkc:XX#key-0",
        "type": "Ed25519VerificationKey2020",
        "controller": "did:ccd:NET:pkc:PK",
        "publicKeyMultibase": "fXX"
      }
    ],
    "authentication": [
      {
        "did:ccd:pkc:XX#key-0"
      }
    ]
  }

Identity Provider DID
---------------------

The goal of the Identity Provider DID is identify a Concordium identity provider.

> An identity provider is an organization, approved by Concordium, that performs off-chain identification of users.

Reviewer 4: ```suggestion
An identity provider (IDP) is an organization, approved by Concordium, that performs off-chain identification of users.
```

Author: @abizjak suggested [here](https://github.com/Concordium/concordium-update-proposals/pull/39/files/35062a110099c3074be5c0151507facf54bc6c83#r1222690694) not use the acronym.
I guess we could use "IDP" if it's recognizable, that is, used in other parts of our documentation.

Reviewer 4: It is written out everywhere in the documentation. But the general rule is that as long as you write it out the first time and follow it in parentheses with the acronym, then it's fine to use the acronym in the rest of the document. But I will leave the final decision to you and Ales.

Reviewer 4: That was why I put in IDP next to this. Because IDP is used elsewhere in the same doc.

Author: > That was why I put in IDP next to this. Because IDP is used elsewhere in the same doc.

@dg-concordium you're right, I thought I expanded the acronym in all places in the document, but it's used in several places, my bad. So, either I spell it out in full everywhere or introduce an acronym.

Author: @dg-concordium I'll add the acronym in parentheses, as you suggested. It's used many times, so it makes sense to have it.

The resulting DID is ``did:ccd:NET:sci:IND:SUBIND`` where ``IND``, ``SUBIND`` are the index and the subindex of the instance.

Public Key Cryptography DID
^^^^^^^^^^^^^^^^^^^^^^^^^^^

A public key cryptography DID can be created by generating a fresh Ed25519 key pair.
The resulting DID is ``did:ccd:NET:pkc:PK`` where ``PK`` is the base16 encoded public key.
These DIDs are not registered on the blockchain.

Identity Provider DID
^^^^^^^^^^^^^^^^^^^^^

Identity providers can be added as a `chain update <https://docs.rs/concordium_base/1.2.0/concordium_base/updates/index.html>`_ transaction of type `UpdateAddIdentityProvider <https://docs.rs/concordium_base/1.2.0/concordium_base/updates/enum.UpdateType.html#variant.UpdateAddIdentityProvider>`_.

Read
----

Account DID
^^^^^^^^^^^

The DID document information for a DID of the form

``did:ccd:NET:acc:ADDR``

can be resolved by looking up an account with address ``ADDR`` on blockchain ``NET``.

Data required to construct the DID document can be acquired by using the gRPC interface command ``GetAccountInfo``.

See the details in the `gRPC v2 documentation`_.

> From the command line, ``concordium-client`` allows to retrieve the data in the following way:

Reviewer 4: ```suggestion
From the command line, ``concordium-client`` allows retrieving the data in the following way:
```

.. code-block:: console

    $concordium-client raw GetAccountInfo ADDR

.. TODO update, once we have a DID resolver

Concordium Credential DID
^^^^^^^^^^^^^^^^^^^^^^^^^

The DID document information for a DID of the form

``did:ccd:NET:cred:CRED``

can be resolved by looking up a credential with ID ``CRED`` on blockchain ``NET``.

Data required to construct the DID document can be acquired by using the same gRPC interface command ``GetAccountInfo`` as for Concordium account DIDs.

.. TODO update, once we have a DID resolver

Smart Contract Instance DID
^^^^^^^^^^^^^^^^^^^^^^^^^^^

The DID document information for a DID of the form

``did:ccd:NET:sci:IND:SUBIND``

can be resolved by looking up a smart contract instance with indices ``IND``, ``SUBIND`` on blockchain ``NET``.
This includes a lookup of the owner's account.

Data required to construct the DID document can be acquired by using the gRPC interface command ``GetInstanceInfo``.

See the details in the `gRPC v2 documentation`_.

> From the command line, ``concordium-client`` allows for retrieving the data in the following way:

Reviewer 4: ```suggestion
From the command line, ``concordium-client`` allows retrieving the data in the following way:
```

.. code-block:: console

  $concordium-client contract show IND

.. TODO update, once we have a DID resolver

.. _sci-reference:

Smart Contract Instance Reference
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*Dereferencing* the smart contract DID reference invokes the specified entrypoint.

Dereferencing a DID reference of the form

``did:ccd:NET:sci:IND:SUBIND/EP[/PAR]``

can be done by using the gRPC interface command ``InvokeInstance``.
The entrypoint is considered a *view*: no state changes are persisted, only the result of the invocation is returned to the caller.
The parameter ``PAR`` is passed to the entrypoint.

> The result of the invocation is the return value produced by the contract or an error, if the invocation failed.

Reviewer 4: ```suggestion
The result of the invocation is the return value produced by the contract or an error if the invocation fails.
```

If the contract contains an embedded schema, then the response is the following:

.. code-block:: json

  {
    "type" : "json",
    "data": "JSON"
  }

Where ``JSON`` is the JSON rendering of the response.

If the contract does not contain an embedded schema, then the response is the following:

.. code-block:: json

  {
    "type" : "raw",
    "data": "BASE16DATA"
  }

Where ``BASE16DATA`` is a base16-encoded return value.

> From the command line, ``concordium-client`` allows for invoking a smart contract instance in the following way:

Reviewer 4: ```suggestion
From the command line, ``concordium-client`` allows invoking a smart contract instance in the following way:
```

The base16 encoding of the ``param.bin`` file corresponds to ``PAR``.

See the details in the `gRPC v2 documentation`_.

.. seealso::

  `Dereferencing a DID URL`_ in the W3C Credentials Community Group draft report.

.. TODO update, once we have a DID resolver

Public Key Cryptography DID
^^^^^^^^^^^^^^^^^^^^^^^^^^^

The DID document corresponding to a DID of the form

``did:ccd:NET:pkc:PK``

can be constructed directly from the DID without any lookup necessary.

.. note::

  The ``NET`` part is optional and currently there is no difference how the documents are generated for different networks.
  In the future, however, the ``vefiricationMethod`` as it specified in :ref:`concordium-did-pkc` might depend on the network.

Identity Provider DID
^^^^^^^^^^^^^^^^^^^^^

The DID document information for a DID of the form

``did:ccd:NET:idp:INDEX``

can be resolved by looking up an identity provider ``INDEX`` on blockchain ``NET``.

Data required to construct the DID document can be acquired by using the gRPC interface command ``GetIdentityProviders``.

See the details in the `gRPC v2 documentation`_.

> From the command line, ``concordium-client`` allows to retrieve the data in the following way:

Reviewer 4: ```suggestion
From the command line, ``concordium-client`` allows retrieving the data in the following way:
```

.. code-block:: console

    $concordium-client raw GetIdentityProviders

Update
------

Account DID
^^^^^^^^^^^

It is possible to update Account DID documents by sending an `update credentials transaction <https://docs.rs/concordium_base/1.2.0/concordium_base/transactions/construct/fn.update_credentials.html>`_.

> This type of transactions allows for adding/removing credentials and changing the signature threshold.

Reviewer 4: ```suggestion
This type of transaction allows for adding/removing credentials and changing the signature threshold.
```

Concordium Credential DID
^^^^^^^^^^^^^^^^^^^^^^^^^

It is possible to update Concordium Credential DID documents by sending an `update credentials keys transaction <https://docs.rs/concordium_base/1.2.0/concordium_base/transactions/send/fn.update_credential_keys.html>`_.

> This type of transactions allows for updating the keys associated with a Concordium credential and the corresponding signature threshold.

Reviewer 4: ```suggestion
This type of transaction allows for updating the keys associated with a Concordium credential and the corresponding signature threshold.
```

Deactivate
----------

Concordium Credential DID
^^^^^^^^^^^^^^^^^^^^^^^^^

A Concordium Credential DID can be deactivated by removing the corresponding credential with an `update credentials transaction <https://docs.rs/concordium_base/1.2.0/concordium_base/transactions/construct/fn.update_credentials.html>`_.

Security Considerations
=======================

> The ``did:ccd`` method is built on top the Concordium blockchain, a public permissionless DLT.

Reviewer 3: ```suggestion
The ``did:ccd`` method is built on top the Concordium blockchain, a public permissionless DL.
```

Security of the DID method reduces to the security of the underlying blockchain protocol.
This concerns attacks such as eavesdropping, replay, message insertion, deletion, modification, denial of service, amplification, and man-in-the-middle.

Parties SHOULD run a full node of the underlying blockchain protocol to ensure that they can read and write securely to the DL.

Authorization is performed by means of digital signature keys.
Leakage of private keys allows an attacker to take control.

> Parties therefore MUST handle private keys with care.

Reviewer 4: ```suggestion
Therefore, parties MUST handle private keys with care.
```

- ``@context`` - the attribute that expresses context information.
- ``id`` - the DID of the account.
- ``verificationMethod`` - the account's verification method.
  It is a nested :ref:`threshold scheme <concordium-did-verification-method>` requiring at ``T`` out of ``M`` credentials to sign; each credential uses its own threshold scheme requiring ``R_i`` out of ``N_i`` keys to sign, where ``i = 1..M``.and ``j = 1..N_i``.
  The credentials are identified by a `DID fragment`_ ``#credential-i``, and the keys in each credentials by ``#key-j-i`` where ``i = 1..M`` and ``j = 1..N_i``.
- ``authentication`` - authentication method for the account.

The document MAY include any other public data of a Concordium account.

.. note::

  A `DID fragment`_ allows for referencing a particular credential, or a key in the Account DID Document.
  The fragment is used to locate the (unique) JSON object by matching the DID URL with the object's ``id`` property.

.. seealso::

  `Dereferencing a DID URL`_ in the W3C Credentials Community Group draft report.

.. code-block:: json

  {
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "Concordium DID URI"

>     ],

Reviewer 2: Are we going to keep these contexts?

Author: Let's drop them for now. The standard admits JSON representation that does not use contexts.

Author: I removed the contexts.

            "verificationMethod": [
              {
                "id": "did:ccd:NET:acc:ADDR#credential-M",
                "controller": "did:ccd:NET:acc:ADDR",
                "type": "VerifiableCondition2021",
                "threshold": "N",
                "conditionThreshold": [
                  {
                    "id": "did:ccd:NET:acc:ADDR#key-1-M",
                    "type": "Ed25519VerificationKey2020",
                    "controller": "did:ccd:NET:acc:ADDR",
                    "publicKeyMultibase": "fVV"
                  },
                  "...",
                  {
                    "id": "did:ccd:NET:acc:ADDR#key-N_M-M",
                    "type": "Ed25519VerificationKey2020",
                    "controller": "did:ccd:NET:acc:ADDR",
                    "publicKeyMultibase": "fZZ"
                  }
                ]
              }
            ]
          }
        ]
      }
    ],
    "authentication": [
      "#acc-1"
    ]
  }

.. note::
  The ``publicKeyMultibase`` field contains a public key prefixed with ``f`` that denotes the base16 encoding.
  See `The Multibase Encoding Scheme`_.

Concordium Credential DID
-------------------------

> The goal of the Account DID Document is to provide information about Concordium credentials, including a possibility to reference particular pieces of data, such as public keys.

Reviewer 2: Shouldn't this talk about Credential DID document instead?

The following document defines a Concordium credential with ID ``CRED``.
The credential has ``N`` keys and uses a threshold signature scheme requiring ``T`` signatures.

.. code-block:: json

  {
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "Concordium DID URI"
    ],
    "id": "did:ccd:NET:cred:CRED#credential-1",
    "verificationMethod": [
      {
        "type": "VerifiableCondition2021",
        "threshold": "T",
        "conditionThreshold": [
          {
            "id": "did:ccd:NET:cred:CRED#key-1",
            "type": "Ed25519VerificationKey2020",
            "controller": "did:ccd:NET:cred:CRED",
            "publicKeyMultibase": "fXX"
          },
          "...",
          {
            "id": "did:ccd:NET:cred:CRED#key-N",
            "type": "Ed25519VerificationKey2020",
            "controller": "did:ccd:NET:cred:CRED",
            "publicKeyMultibase": "fYY"
          }
        ]
      }
    ],
    "authentication": [
      "#credential-1"
    ]
  }

Smart Contract Instance DID
---------------------------

The goal of the Smart Contract Instance DID is to provide meta-data about the contract instance.

> At the moment, the main piece of meta-data is the Concordium account that send the initialization transaction.

Reviewer 2: Probably the more important part is the list of entrypoints.

Author: Fixed.
